### PR TITLE
Inline ClassType, SelfTypeParam, AliasType, SelfType, LiteralType, BlamedUntyped into TypePtr.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Inline ClassType, SelfTypeParam, AliasType, SelfType, LiteralType, BlamedUntyped into TypePtr.

Inlined types have the following restrictions:
* You _cannot_ use `cast_type` on them, since `cast_type` returns a pointer. There is a static_assert that tells you to use `isa_type` to check the type within the TypePtr and `cast_type_nonnull` to cast a TypePtr to a type.
  * 99% of this diff is converting `cast_type` to `isa_type`/`cast_type_nonnull` checks.

Making a new inlined type involves the following steps (compilation will fail if any are missed :)):
* The type must be defined with `TYPE_INLINED`.
* `make_type` must contain specializations for every unique combination of parameter types passed to `make_type<YourClass>`.
  * This is a limitation of C++, which does not support partial specialization of function templates.
* `cast_type_nonnull` must contain a specialization for your class that returns a value type rather than a reference type.

Typecase still supports lambdas that accept inlined types as const references, as C++ allows local const references to prolong the lifetime of values. Thus, I did not have to change the type signature of any typecase lambdas! 🎉  

### Motivation

Speed!

* Inlined types no longer have to increment/decrement an atomic reference counter, which is a scalability bottleneck for commonly accessed types (e.g., nilType).
* Inlined types do not have to allocate a Type object or an atomic counter object on the heap.
* Inlined types can be compared by value.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_INx7zjNvIt51AS

File: core/TypePtr.h [hunk 1]
```diff
@@ -131,7 +206,11 @@ class TypePtr final {
 
     TypePtr &operator=(TypePtr &&other) noexcept {
         handleDelete();
-        counter = other.releaseCounter();
+        if (other.containsPtr()) {
```
Comment (elliottt) [inline core/TypePtr.h:209]: What about when `*this == other`?

File: core/Types.h [hunk 2]
```diff
@@ -312,6 +331,24 @@ TYPE(ClassType) {
 };
 CheckSize(ClassType, 8, 8);
 
+template <> inline TypePtr make_type<ClassType, core::SymbolRef &>(core::SymbolRef &ref) {
+    return TypePtr(TypePtr::Tag::ClassType, ref.rawId(), 0);
+}
+
+template <> inline TypePtr make_type<ClassType, core::SymbolRef>(core::SymbolRef &&ref) {
+    return TypePtr(TypePtr::Tag::ClassType, ref.rawId(), 0);
+}
```
Comment (elliottt) [inline core/Types.h:340]: Just golfing on memory layout, but: what if you store the rawId in `value` instead? It's not used for anything at this point, and if the tags for inlined values always fit in the first three bits of the ptr, you could do a pretty quick check and then return `value`, avoiding the shift. You could also use the first three bits being non-null for the inlined check this way.

**Review by elliottt — APPROVED** [review]
This looks great! Super excited for the perf improvements 🎉 

I wondered about maybe always storing the values of inlined types in `value`, so that you could avoid the shift to retrieve it, but it's definitely a non-blocking change.

File: core/Types.h [hunk 3]
```diff
@@ -64,6 +65,8 @@ class ArgInfo {
 CheckSize(ArgInfo, 48, 8);
 
 template <class T, class... Args> TypePtr make_type(Args &&... args) {
+    // Every inlined type must specialize `make_type`.
+    static_assert(!TypePtr::TypeToIsInlined<T>::value);
```
[the remark this answers is not in the document]
Comment (jvilk-stripe) [reply core/Types.h:69]: Can do.

File: core/Types.h (hunk 2 above)
Comment (jvilk-stripe) [reply core/Types.h:340]: I can explore that as a follow-up change and see if there's any perf change. I suspect that, with everything being in cache, there's no difference.

**Pull request merged**